src/lit-html.ts

```diff
@@ -220,10 +221,18 @@ export class Template {
             parent.insertBefore(document.createTextNode(strings[i]), node);
             this.parts.push(new TemplatePart('node', index++));
           }
-        } else if (!node.nodeValue!.trim()) {
-          nodesToRemove.push(node);
-          currentNode = previousNode;
-          index--;
+        } else {
```

Comment:
These conditions are starting to become dense, rendering the `while` loop difficult to consider in its entirety. Consider factoring these conditional paths into methods for readability.

Reply:
+1

Reply:
The difficulty in splitting out methods is the amount of mutable state they share. It would require either a lot of parameters and return values, plus assignments to the shared state, or splitting into closures within the constructor, which might create more closure allocation per template.

I'd like to take a look at refactoring separately because of the amount of changes it'd require.